sdk/src/extras/disk.rs

```diff
@@ -328,56 +325,55 @@ pub mod types {
                     DiskImportError::context("starting the build write process failed", e)
                 })?;
 
-            // Create one tokio task for each thread that will upload file chunks
             let mut handles: Vec<tokio::task::JoinHandle<Result<(), DiskImportError>>> =
-                Vec::with_capacity(self.upload_thread_ct);
-            let mut senders = Vec::with_capacity(self.upload_thread_ct);
-
-            for _ in 0..self.upload_thread_ct {
-                let (tx, mut rx) = mpsc::channel(100);
+                Vec::with_capacity(self.upload_task_ct.get());
+            let (tx, rx) = flume::bounded(64);
+            let (failed_tx, failed_rx) = flume::bounded(self.upload_task_ct.get());
+            let (resubmit_tx, resubmit_rx) = flume::bounded(self.upload_task_ct.get());
+
+            for _ in 0..self.upload_task_ct.get() {
+                let mut worker = UploadWorker {
+                    client: self.client.clone(),
+                    disk: self.disk.clone(),
+                    project: self.project.clone(),
+                    progress_tx: self.progress_tx.clone(),
+                };
 
-                let client = self.client.clone();
-                let project = self.project.clone();
-                let disk = self.disk.clone();
-                let upload_progress = self.upload_progress.clone();
-                let progress_tx = self.progress_tx.clone();
+                let rx = rx.clone();
+                let failed_tx = failed_tx.clone();
+                let resubmit_rx = resubmit_rx.clone();
 
                 handles.push(tokio::spawn(async move {
-                    while let Some((offset, base64_encoded_data, data_len)) = rx.recv().await {
-                        client
-                            .disk_bulk_write_import()
-                            .disk(disk.clone())
-                            .project(project.clone())
-                            .body(ImportBlocksBulkWrite {
-                                offset,
-                                base64_encoded_data,
-                            })
-                            .send()
-                            .await?;
-
-                        upload_progress.fetch_add(data_len, Ordering::Relaxed);
-                        progress_tx.send_replace(upload_progress.load(Ordering::Relaxed));
+                    while let Ok(chunk) = rx.recv_async().await {
+                        if let Err(e) = worker.upload_chunk(&chunk).await {
+                            let _ = failed_tx.send_async(chunk).await;
+                            return Err(e);
+                        }
+                    }
+
+                    // Signal the main task that the worker has sent all chunks successfully and
+                    // none remain to be sent.
+                    drop(failed_tx);
+
+                    while let Ok(chunk) = resubmit_rx.recv_async().await {
+                        worker.upload_chunk(&chunk).await?;
```

Comment:
why are we not putting failures back on the failed_tx?

Reply:
This is something I went back and forth on.

Workers close `failed_tx` to indicate they has finished all requests in `tx`. If we keep `failed_rx` open for resubmits then we deadlock.

We could add yet another channel for failed resubmits, but given that we would have at most <WORKER_CT - 1> retries, I didn't think it was worth the extra complexity.

Reply:
let's not deadlock! this is already better than it was so let's just proceed